### PR TITLE
Fix: distributed sampler for multi GPU

### LINKED ISSUE #3047: Multi-GPU fails when using ConcatDataLoader
When utilizing multi-gpu training with DDP, e.g. with a `SemiSupervisedDataLoader` and subsequently `ConcatDataLoader`, an error is raised because `distributed_sampler` keyword argument required for `AnnDataLoader` is not explicit in the class definition of `ConcatDataLoader`.  Passing this keyword argument in using the wildcard allows `AnnDataLoader` to be initialized properly, but throws an error for initializing the parent class `torch.util.DataLoader` using the same `distributed_sampler` keyword argument.

```python
datasplitter_kwargs = {}
datasplitter_kwargs['distributed_sampler'] = True
data_splitter = SemiSupervisedDataSplitter(
            adata_manager=self.adata_manager,
            train_size=train_size,
            validation_size=validation_size,
            shuffle_set_split=shuffle_set_split,
            n_samples_per_label=n_samples_per_label,
            batch_size=batch_size,
            **datasplitter_kwargs,
        )
```

```pytb
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/scvi/train/_trainrunner.py", line 98, in __call__
2024-11-18T23:06:15.416Z [rank1]:     self.trainer.fit(self.training_plan, self.data_splitter)
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/scvi/train/_trainer.py", line 219, in fit
2024-11-18T23:06:15.416Z [rank1]:     super().fit(*args, **kwargs)
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/pytorch/trainer/trainer.py", line 544, in fit
2024-11-18T23:06:15.416Z [rank1]:     call._call_and_handle_interrupt(
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/pytorch/trainer/call.py", line 43, in _call_and_handle_interrupt
2024-11-18T23:06:15.416Z [rank1]:     return trainer.strategy.launcher.launch(trainer_fn, *args, trainer=trainer, **kwargs)
2024-11-18T23:06:15.416Z [rank1]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/pytorch/strategies/launchers/subprocess_script.py", line 102, in launch
2024-11-18T23:06:15.416Z [rank1]:     return function(*args, **kwargs)
2024-11-18T23:06:15.416Z [rank1]:            ^^^^^^^^^^^^^^^^^^^^^^^^^
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/pytorch/trainer/trainer.py", line 580, in _fit_impl
2024-11-18T23:06:15.416Z [rank1]:     self._run(model, ckpt_path=ckpt_path)
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/pytorch/trainer/trainer.py", line 989, in _run
2024-11-18T23:06:15.416Z [rank1]:     results = self._run_stage()
2024-11-18T23:06:15.416Z [rank1]:               ^^^^^^^^^^^^^^^^^
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/pytorch/trainer/trainer.py", line 1035, in _run_stage
2024-11-18T23:06:15.416Z [rank1]:     self.fit_loop.run()
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/pytorch/loops/fit_loop.py", line 194, in run
2024-11-18T23:06:15.416Z [rank1]:     self.setup_data()
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/pytorch/loops/fit_loop.py", line 222, in setup_data
2024-11-18T23:06:15.416Z [rank1]:     train_dataloader = _request_dataloader(source)
2024-11-18T23:06:15.416Z [rank1]:                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/pytorch/trainer/connectors/data_connector.py", line 342, in _request_dataloader
2024-11-18T23:06:15.416Z [rank1]:     return data_source.dataloader()
2024-11-18T23:06:15.416Z [rank1]:            ^^^^^^^^^^^^^^^^^^^^^^^^
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/pytorch/trainer/connectors/data_connector.py", line 309, in dataloader
2024-11-18T23:06:15.416Z [rank1]:     return call._call_lightning_datamodule_hook(self.instance.trainer, self.name)
2024-11-18T23:06:15.416Z [rank1]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/pytorch/trainer/call.py", line 179, in _call_lightning_datamodule_hook
2024-11-18T23:06:15.416Z [rank1]:     return fn(*args, **kwargs)
2024-11-18T23:06:15.416Z [rank1]:            ^^^^^^^^^^^^^^^^^^^
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/scvi/dataloaders/_data_splitting.py", line 332, in train_dataloader
2024-11-18T23:06:15.416Z [rank1]:     return self.data_loader_class(
2024-11-18T23:06:15.416Z [rank1]:            ^^^^^^^^^^^^^^^^^^^^^^^
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/fabric/utilities/data.py", line 324, in wrapper
2024-11-18T23:06:15.416Z [rank1]:     init(obj, *args, **kwargs)
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/scvi/dataloaders/_semi_dataloader.py", line 75, in __init__
2024-11-18T23:06:15.416Z [rank1]:     super().__init__(
2024-11-18T23:06:15.416Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/fabric/utilities/data.py", line 324, in wrapper
2024-11-18T23:06:15.417Z [rank1]:     init(obj, *args, **kwargs)
2024-11-18T23:06:15.417Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/scvi/dataloaders/_concat_dataloader.py", line 65, in __init__
2024-11-18T23:06:15.417Z [rank1]:     super().__init__(self.largest_dl, **data_loader_kwargs)
2024-11-18T23:06:15.417Z [rank1]:   File "/stage/env/lib/python3.11/site-packages/lightning/fabric/utilities/data.py", line 324, in wrapper
2024-11-18T23:06:15.417Z [rank1]:     init(obj, *args, **kwargs)
2024-11-18T23:06:15.417Z [rank1]: TypeError: DataLoader.__init__() got an unexpected keyword argument 'distributed_sampler'
```

#### Versions:

> VERSION 1.1.2

### PR DESCRIPTION
**Pull request by ori-kron-wis**
Update for the dataloader in case of distributed sampler

close https://github.com/scverse/scvi-tools/issues/3047

### CODE REVIEW

File: tests/dataloaders/test_dataloaders.py [hunk 1]
```diff
@@ -92,29 +95,74 @@ def test_anndataloader_distributed_sampler_init():
...
+    rank: int,
+    world_size: int,
+    manager: scvi.data.AnnDataManager,
+    save_path: str,
+    datasplitter_kwargs,
 ):
     # initializes the distributed backend that takes care of synchronizing processes
     torch.distributed.init_process_group(
-        "gloo",  # backend that works on all systems
+        "nccl",  # backend that works on all systems
```
Comment (canergen) [inline tests/dataloaders/test_dataloaders.py:106]: Remove the comment that it works on all systems

File: tests/dataloaders/test_dataloaders.py [hunk 2]
```diff
@@ -92,29 +95,74 @@ def test_anndataloader_distributed_sampler_init():
...
     return
 
 
-@pytest.mark.optional
-def test_anndataloader_distributed_sampler(save_path: str, num_processes: int = 2):
-    adata = scvi.data.synthetic_iid()
-    manager = generic_setup_adata_manager(adata)
+@pytest.mark.parametrize("num_processes", [1, 2])
+def test_anndataloader_distributed_sampler(num_processes: int, save_path: str):
+    if torch.cuda.is_available():
```
Comment (canergen) [inline tests/dataloaders/test_dataloaders.py:120]: Shouldn't it be assert here?

File: tests/model/test_scanvi.py [hunk 3]
```diff
@@ -578,3 +578,63 @@ def check_no_logits_and_softmax(model: SCANVI):
 
     model = SCANVI.load(resave_model_path, adata)
     check_no_logits_and_softmax(model)
+
```
Comment (canergen) [inline tests/model/test_scanvi.py:581]: Shouldn't this be in another test all together?

File: tests/model/test_scvi.py [hunk 4]
```diff
@@ -1297,3 +1297,57 @@ def test_scvi_num_workers():
     model.get_reconstruction_error()
     model.get_normalized_expression(transform_batch="batch_1")
     model.get_normalized_expression(n_samples=2)
+
```
Comment (canergen) [inline tests/model/test_scvi.py:1300]: Same here. Just have one test for ddp and not in the test_scvi thing.

File: src/scvi/dataloaders/_samplers.py [hunk 5]
```diff
@@ -36,6 +36,14 @@ def __init__(
         drop_dataset_tail: bool = False,
         **kwargs,
     ):
+        # for redundant_key in [
```
Comment (canergen) [inline src/scvi/dataloaders/_samplers.py:39]: Still necessary?

File: src/scvi/dataloaders/_concat_dataloader.py [hunk 6]
```diff
@@ -45,6 +53,12 @@ def __init__(
         self._shuffle = shuffle
         self._batch_size = batch_size
         self._drop_last = drop_last
+        self._distributed_sampler = distributed_sampler
+        # self._drop_dataset_tail = (
```
Comment (canergen) [inline src/scvi/dataloaders/_concat_dataloader.py:57]: Do we still need this? Drop_last should have the same effect?

**Review by canergen — APPROVED** [review]
LGTM. See minor comments.

File: tests/dataloaders/test_dataloaders.py (hunk 1 above)
Comment (ori-kron-wis) [reply tests/dataloaders/test_dataloaders.py:106]: fixed

File: tests/model/test_scanvi.py (hunk 3 above)
Comment (ori-kron-wis) [reply tests/model/test_scanvi.py:581]: ok

File: tests/model/test_scvi.py (hunk 4 above)
Comment (ori-kron-wis) [reply tests/model/test_scvi.py:1300]: ok

File: tests/dataloaders/test_dataloaders.py (hunk 2 above)
Comment (ori-kron-wis) [reply tests/dataloaders/test_dataloaders.py:120]: So, there is no such thing as a CUDA flag and I dont want the test to fail on a CPU run.

File: tests/model/test_scvi.py (hunk 4 above)
Comment (ori-kron-wis) [reply tests/model/test_scvi.py]: we can add a label multi-gpu to test those kind of tests

**Pull request merged**